nethack_server/src/db_pgsql.c

```diff
@@ -686,4 +763,83 @@ db_add_topten_entry(int gid, int points, int hp, int maxhp, int deaths,
     return;
 }
 
+struct nh_board_entry *
+db_get_board_entries(char *since, int *length) {
+    PGresult *res;
+    int i, levelcol, depthcol, hpcol, hpmaxcol, encol, enmaxcol, wicol, incol,
+        stcol, dxcol, cocol, chcol, tscol, movescol, plrolecol, plracecol,
+        plgendcol, plaligncol, namecol, leveldesccol;
+    const char *plrole, *plrace, *plgend, *plalign, *name, *leveldesc, *ts;
+
+    const char *const params[] = { since };
+    const int paramFormats[] = { 0 };
+    struct nh_board_entry *entries;
+
+    res =
+        PQexecParams(conn, SQL_get_games_since, 1, NULL, params, NULL,
+                     paramFormats, 0);
+    if(PQresultStatus(res) != PGRES_TUPLES_OK) {
+        log_msg("db_get_board_entries error: %s", PQerrorMessage(conn));
+        PQclear(res);
+        return NULL;
+    }
+    *length = PQntuples(res);
+    levelcol = PQfnumber(res, "level");
+    depthcol = PQfnumber(res, "depth");
+    hpcol = PQfnumber(res, "health");
+    hpmaxcol = PQfnumber(res, "maxhealth");
+    encol = PQfnumber(res, "energy");
+    enmaxcol = PQfnumber(res, "maxenergy");
+    wicol = PQfnumber(res, "wi");
+    incol = PQfnumber(res, "intel");
+    stcol = PQfnumber(res, "str");
+    dxcol = PQfnumber(res, "dx");
+    cocol = PQfnumber(res, "co");
+    chcol = PQfnumber(res, "ch");
+    tscol = PQfnumber(res, "ts");
+    movescol = PQfnumber(res, "moves");
+    plrolecol = PQfnumber(res, "role");
+    plracecol = PQfnumber(res, "race");
+    plgendcol = PQfnumber(res, "gender");
+    plaligncol = PQfnumber(res, "alignment");
+    namecol = PQfnumber(res, "plname");
+    leveldesccol = PQfnumber(res, "level_desc");
+
+    entries = malloc(sizeof (struct nh_board_entry) * (*length));
+    for (i = 0; i < *length; i++) {
+        entries[i].level = atoi(PQgetvalue(res, i, levelcol));
+        entries[i].depth = atoi(PQgetvalue(res, i, depthcol));
+        entries[i].hp = atoi(PQgetvalue(res, i, hpcol));
+        entries[i].hpmax = atoi(PQgetvalue(res, i, hpmaxcol));
+        entries[i].en = atoi(PQgetvalue(res, i, encol));
+        entries[i].enmax = atoi(PQgetvalue(res, i, enmaxcol));
+        entries[i].wi = atoi(PQgetvalue(res, i, wicol));
+        entries[i].in = atoi(PQgetvalue(res, i, incol));
+        entries[i].st = atoi(PQgetvalue(res, i, stcol));
+        entries[i].dx = atoi(PQgetvalue(res, i, dxcol));
+        entries[i].co = atoi(PQgetvalue(res, i, cocol));
+        entries[i].ch = atoi(PQgetvalue(res, i, chcol));
+        entries[i].moves = atoi(PQgetvalue(res, i, movescol));
```

Comment:
I'm not familiar with the schema here, but since postgres types generally don't align with c types, it's dangerous to apply `atoi(3)` here. Import `strtonum(3)` from OpenBSD and use that, or settle with `strtoll(3)` and check for overflow.